New Graph patterns (#336)
Microsoft/api-guidelines: graph/patterns/namespace.md

> # Namespace

Reviewer: Or should we call this "Type namespaces"

Microsoft Graph API Design Pattern

### *The Namespace provides the ability to organize resource definitions together into a logical set.*

## Problem

When building a complex offering API designers may need to model many different
resources and their relationships. For better user experience and

> discoverability related API elements need to be grouped together.  

Reviewer: elements or types?

Author: elements will include types and operations

## Solution

API designers can use the Namespace attribute of the CSDL schema to declare a
namespace and logically organize related API entities in the Graph metadata.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ XML
<Schema Namespace="microsoft.graph.{namespace}">
...
</Schema>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> A public namespace must have "microsoft.graph.” prefix and be presented in camel

Reviewer: Nit:  different quote marks :)

case, i.e microsoft.graph.myNamespace.

When type casting is required in the API query, request or response, a fully
qualified type name is represented as concatenation of a namespace and a type

> name. For consistent user experience namespaces should be aligned with the corresponding API category path segment.

Reviewer 2: ```suggestion
name. For consistent user experience namespaces MUST be aligned with the corresponding API category path segment.
```

## When to Use this Pattern

API resource grouping creates a user-friendly experience keeping all resources

> for a specific feature close together. It also allows generating smallersmaller downloadable client components 

Reviewer 3: ```suggestion
for a specific feature close together. It also allows generating smaller downloadable client components 
```

> for a specific feature close together. It also allows generating smallersmaller downloadable client components 

Reviewer 2: ```suggestion
for a specific feature close together
```
This isn't true due to circular dependencies

> for a specific feature close together. It also allows generating smallersmaller downloadable client components 

Reviewer: Nit: double smaller

> for a specific feature close together. It also allows generating smallersmaller downloadable client components 

Reviewer: @darrelmiller should we be mentioning smaller downloadable client components?  Is that really true?

> and limits the length of IDE prompts such as Intellisense.

Reviewer 2: ```suggestion
and limits the length of IDE prompts such as auto-complete in some programming languages.
```

> We recommend creating a new namespace when a new top-level API category is

Reviewer 2: As discussed, I am not convinced this is what we should recommend.

Author: changed to aligned as discussed at API Council

introduced.

> ## Issues and Considerations

Reviewer: You should call out that only one level of nesting deeper than microsoft.graph is supported, and that if you define something more deeply nested, the type will be coerced into the 3rd level namespace.

Is that implementation detail?

Author: added as a constraint for clarity

1.  Microsoft Graph consistency requirements discourage using the same type
    names for different concepts even within different namespaces. Microsoft
    Graph type names must be descriptive and unique within the API
    surface without requiring  full qualification.

> 2.  Namespace must be consistent with API category in the navigation path according to [Microsoft Graph REST API Guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/graph/GuidelinesGraph.md#uniform-resource-locators-urls)

Reviewer 3: Should me have this a relative URL so the link won't break in a different branch?
```suggestion
2.  Namespace must be consistent with API category in the navigation path according to [Microsoft Graph REST API Guidelines](GuidelinesGraph.md#uniform-resource-locators-urls)
```

3.  When type name is ambiguous and requires a namespace qualifier, changing
    namespace is a breaking change.

> 4.  To extend a type in a different schema, a service must declare that schema

Reviewer 2: This seems like internal implementation details.

Author: it doesn't look like AGS specific consideration and is used quite often for graph.entity. <EntityType Name="callRecord" BaseType="graph.entity"

    and the type in it. This is conceptually similar to .NET partial types.

5.  To reference a type in a different schema, simply refer to that type by
    fully qualified name (namespace + type name).

6.  Cyclical references between namespaces are not allowed as many

>     object-oriented languages don’t support a cycle in inheritance.

Reviewer 2: ```suggestion
    object-oriented languages don’t support a cycles between namespaces.
```
The constraint is not related to specifically to inheritance

7.  Microsoft Graph has heuristic rules for declared namespaces:

    1.  All public namespaces must have a prefix ‘microsoft.graph’

    2.  If a namespace does not begin with ‘microsoft.graph’ prefix, all types
        in the schema will be coerced into the main ‘microsoft.graph’ namespace.

## Examples

### Namespace and type declarations:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ XML
<Schema Namespace="microsoft.graph.search" xmlns=”<http://docs.oasis-open.org/odata/ns/edm>”\>
…
    <EntityType Name="bookmark" …
    </EntityType>
…
</Schema>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Fully qualified type name: microsoft.graph.search.bookmark

### Managing multiple schemas:

Workloads must define schemas in their csdl using the Edmx format.

> [Microsoft.IC3.DataPlatform](https://dev.azure.com/msazure/One/_git/AD-AggregatorService-Workloads?path=%2FWorkloads%2FMicrosoft.IC3.DataPlatform%2Foverride%2Fschema-Prod-beta.csdl)

Reviewer: This doesn't seem like this should be in the public docs.
